src/twisted/trial/_asynctest.py

```diff
@@ -199,7 +199,8 @@ def deferRunCleanups(self, ignored, result):
         object.
         """
         failures = []
-        for func, args, kwargs in self._cleanups[::-1]:
+        while len(self._cleanups) > 0:
+            func, args, kwargs = self._cleanups.pop()
```

Comment:
maybe `consume(func, iter)`...? or `consumeIterator(func, iter)` .. or maybe the args go the other way, I didn't do a survey.

Reply:
I'd probably lean towards a refactoring that avoids the need to mutate state wherever possible.  Since mutation is used heavily throughout trial and unittest, this is probably at least a huge amount of work, though.

It's not obvious to me how `destructivelyIterate` would work with the order reversal here.  I don't think you could make `destructivelyIterate(self_cleanups[::-1])` work.  I suppose you could always add more parameters, `destructivelyIterate(self._cleanups, reverse=True)` or something.  Maybe if `_cleanups` were a stack instead of a list it would go a bit more smoothly.  I haven't looked over the rest of the implementation to see what other trouble such a change might make.

Likewise I'm not sure how `consume(func, iter(self._cleanups))` would work.  If a function is going to help, I think it needs the list object so it can pop items off of it.  (If passing in the list itself was the idea then it seems like it ends up equivalent to `destructivelyIterate`?  But maybe I'm overlooking something.)